Core/GDCore/Events/CodeGeneration/EventsCodeGenerator.h

```diff
@@ -155,7 +155,10 @@ class GD_CORE_API EventsCodeGenerator {
    * \return Code
    */
   gd::String GenerateActionCode(gd::Instruction& action,
-                                EventsCodeGenerationContext& context);
+                                EventsCodeGenerationContext& context,
+                                // TODO: remove this in favor of a `GenerateStatementCode(GenerateActionCode(...))` which simply outputs a ;

```

Comment:
@AlexandreSi even tho this is NOT my code nor my PR I find this comment a bit aggressive, I think that async. events and wait actions are very hard to code and asking when it should be done is rude because arthuro is doing this on he's free time.

GDevelop is an open source project, every contributors is here because of their willing. I, with arthuro and 2 other friends participated on the first GDevelop game jam this week and the week prior to this one, this is why arthuro did not have the time to work on this PR.

Note that arthuro does NOT endorse what I'm saying and I am just sharing my thoughts.
Also please don't take this comment as something rude but more like a request because I think that arthuro should not be pressured.

Reply:
@Oxey405 Your comment is misplaced and misjuding the intentions, there is no aggressivity here at all. Alexandre is part of the GDevelop team and doing a review so that we can identify what needs to be done before merging this PR.
His comment is just asking if the TODO (that I added myself) must be addressed before this PR is merged or if this is a TODO for alter.

Reply:
To come back to this: this should be addressed before this PR is merged, to ensure the code generation keeps a style where each function generate a code, but is not given some partial prefix/suffix code which make it harder to know if the generated code will be correct (and this is less "AST" style as an approach).

Reply:
I'm not sure what this `GenerateStatementCode` would do. Do you mean just appending `";"` to the end of its parameter? That wouldn't work, since non-free actions are not just method calls, they are inside a for loop. Unless you mean `GenerateStatementCode` would also generate the for loops for non-free actions? In that case, the name wouldn't be very fitting in my opinion...

Reply:
> Do you mean just appending ";" to the end of its parameter?

Yep - it's taking some code generated (the parameter) and transform it in a statement by adding a `;` at the end: `GenerateStatementCode(GenerateActionCode(...)) `

> That wouldn't work, since non-free actions are not just method calls, they are inside a for loop

In this case, the caller would just not call GenerateStatementCode: `GenerateActionCode(...))` as before.

But this discussion is actually becoming useless because I see that you removed the default prefix and suffix, they are both now empty strings :) 
So you can remove them entirely (i.e: return to what we had before), remove this TODO, and finally address the other related TODO:

```
        // TODO: rework this into a `codeGenerator.GenerateStoreTaskCode(codeGenerator.GenerateActionCode(...))`
        const gd::String taskSchedulingCode = codeGenerator.GenerateActionCode(
            event.GetInstruction(),
            parentContext,
            "runtimeScene.getAsyncTasksManager().addTask(",
            ", " + callbackCallCode + ");\n");
```

by creating a new "GenerateStoreTaskCode" in the events generator that does something like `return "runtimeScene.getAsyncTasksManager().addTask(" + taskCode + ", " + callbackCallCode + ");\n"`.

This means that we keep the JS code generation as much as possible in the code generator itself. The consumer (like the code generator of `BuiltinCommonInstructions::Async`) are, as much as possible, only calling functions, a bit like they would represent nodes in an abstract syntax tree:

`codeGenerator.GenerateStoreTaskCode(codeGenerator.GenerateActionCode(...), callbackCallCode)`

All of this to say basically to put the prefix/suffix in a separate function 😄

Reply:
> @Oxey405 Your comment is misplaced and misjuding the intentions, there is no aggressivity here at all. Alexandre is part of the GDevelop team and doing a review so that we can identify what needs to be done before merging this PR. His comment is just asking if the TODO (that I added myself) must be addressed before this PR is merged or if this is a TODO for later after the PR is merged.

understood I remove it

Reply:
> by creating a new "GenerateStoreTaskCode" in the events generator that does something like return "runtimeScene.getAsyncTasksManager().addTask(" + taskCode + ", " + callbackCallCode + ");\n".
> [...]
> `codeGenerator.GenerateStoreTaskCode(codeGenerator.GenerateActionCode(...), callbackCallCode)`

Again, that wouldn't work, because if an object action is not an expression when generated as an object action. If an action accepts an object as parameter, this will generate something like this:
```js
runtimeScene.getAsyncTasksManager().addTask(
  for(const i of objectsList1234) {objectsList[1234].myAsyncAction();}
```
... which obviously wouldn't work. The prefix is there so that we can tell the object actions and behaviors action generator to add them not at the beginning and the end of the function result, but if the actual action call, which is in the middle of that code:
```js
for(const i of objectsList1234) {
  runtimeScene.getAsyncTasksManager().addTask(objectsList[1234].myAsyncAction(), asyncCallback1234);
}
```

Reply:
Sorry for this confusion on my side, but this actually is revealing a problem:
How do we handle the continuation of async actions for object actions?

Turns out we never discussed about it. While your solution works, it means that it will run the subsequent event once per object (so if you have 1000 objects, and then in a subevent a collision condition, you will do this collision condition 1000 times).

Can we clarify how this should be handled in the main discussion?